src/v/cloud_storage/remote_segment.cc

```diff
@@ -584,7 +584,7 @@ ss::future<> remote_segment::run_hydrate_bg() {
               _wait_list.size(),
               _data_file ? "available" : "not available");
             std::exception_ptr err;
-            if (!_data_file) {
+            if (!_data_file || !_tx_range) {
```

Comment:
I think the condition is wrong. It should be && because we need both. Also, it is possible for `_tx_range` to be `std::nullopt` when the tx manifest does not exist. I think that you need to make change to `do_materialize_txrange` to always set this value. Currently, if the segment is compacted we're filling this value by calling `_tx_range.emplace()` but if the manifest is absent we're keeping it `std::nullopt`. In order to use it in the condition you need to fix this, I think.

Reply:
> It should be && because we need both.

I think the condition is correct. We attempt to hydrate if `_tx_range` or the `_data_file` have not been set.

> I think that you need to make change to do_materialize_txrange to always set this value. 

I thought so too, but `do_hydrate_tx_range` sets `_tx_range` to an empty vector even when it receives
a not found response. Isn't this enough?